### PR TITLE
Fix object radio

### PR DESCRIPTION
**Pull request by BryceStevenWilley**
The values of object radio buttons are variable names encoded as base64.
Convert those values into the human readable versions of themselves if they can
be converted.

* fromBase64 would accept invalid base64 strings, and just convert it to what it
  wanted to. This does a rough check to make sure that the value contains only
  valid base64 characters before trying to convert it. TODO: there could be more
  work done here to only add values that once decoded, end up as ascii
  characters (since we're working in python), but that can be future work.
* Added more docstrings to the toBase64 and fromBase64 functions
* Made unit tests pass by removing bad base64 conversions from the fixtures
* switched one of the test argument orders, since the expected and actual values
  are backwards. TODO: should change the other as well
  
Tests pass locally (and in the second commit), and don't pass even when the interview steps are kept, meaning 
that the fix does actually fix a DA related bug. The third commit (last one as of writing) fails because it includes extra steps not yet merged in ALAutomatedTestsingTests.

### CODE REVIEW

**Comment by BryceStevenWilley** [comment]
TODO:
* [ ] make a test for checkboxes, specifically with a weird unicode string as a choice and make sure it's choosable
* [ ] make a new docassemble YAML file for the above test (the "decoding" test)

**Comment by plocket** [comment]
Some of the fixtures are failing. A bit braindead to dig into why right now, sorry. Fixtures are sadly fragile. I hesitated to use them in the first place. It seemed like a shame not to unit test them when the output was a string, though. We might discuss removing them.

**Comment by BryceStevenWilley** [comment]
I am aware and know why, see https://github.com/SuffolkLITLab/ALKiln/pull/581/commits/f81a49fb4cb4e56efcc36731da3f3ee5f8be44bb.  

> unit tests are broken because I added `multiple` as an attribute to the fields, so now the deep equals are wrong

Was also having issues with chai not actually printing the diff at all (it just says the assert failed and doesn't print anything else), which was pretty annoying. 

In general I don't have the bandwidth to work on this for a bit. This was supposed to be a smaller commit (fixing just object radio buttons), but has ballooned in scope to try to handle all of the base64 fields, as well as handling object multiselects and various other fields that I don't have the experience with.

**Comment by BryceStevenWilley** [comment]
Since we mentioned (offline) that we should still include some of the changes from #604 and this PR in the stripped down PR, it's just easier to take out the code trying to handle multiples. That code is now in the `fix_object_multiselect` branch (https://github.com/SuffolkLITLab/ALKiln/pull/new/fix_object_multiselect).

I just discovered that I can't run puppeteer in WSL, so unfortunately I can't get things testing and running fully. I anticipate that i'll have to comment out all lines in `test_decoding.yml` that aren't `object_radio` for the tests to fully work, but I can confirm that the fixtures are fine now. Next week I'll have a better setup and will be able to test this fully and get to a point where it's mergable.

**Comment by BryceStevenWilley** [comment]
I've got tests working on this, with the changes in https://github.com/plocket/docassemble-ALAutomatedTestingTests/pull/197.

Should be ready for a last review, @plocket

File: lib/scope.js [hunk 1]
```diff
@@ -54,7 +54,33 @@ let is_secret_var = function(row) {
...
+let fromBase64 = function(base64_str) {
+  /* Takes a base64 encoded string and returns it decoded.
+   * Note that this accept take strings with invalid base 64 characters,
+   * changes those characters to `/`s instead.
+   */
+  let buffer = Buffer.from( base64_str, 'base64' );
+  return buffer.toString( `utf-8` );
+}
+
+let toBase64 = function ( { utf8_str }, exclude_padding=true) {
```
Comment (plocket) [inline lib/scope.js]: Nothing terribly wrong, but the function trigger of `toBase64()` is different than `fromBase64()` in a complex way. The "from" takes a string as an argument, the "to" takes an object with a string as an argument. It might be good to make those more similar for less confusion.

File: lib/scope.js [hunk 2]
```diff
@@ -1200,96 +1234,65 @@ module.exports = {
       // TODO: Should we move this back into `getAllFields()`?
     }
 
-    // A lot of checkbox fields have a key hidden inside them. (not yesno type checkboxes)
-    let checkbox_match = layer_1_decoded.match( /^(.*)\[[BR]'(.*)'\]/ );
+    return scope.nestedDecode(scope, { to_handle: layer_1_decoded, field_like_names, nota });
+  },  // Ends scope.getPossibleVarNames()
+
+  nestedDecode: function ( scope, { to_handle, field_like_names = {}, nota } ) {
+    // A some fields have a key hidden inside them. (not yesno type checkboxes)
```
Comment (plocket) [inline lib/scope.js]: Not crucial
```suggestion
    // Some fields have a key hidden inside them. (not yesno type checkboxes)
```

Comment (plocket) [reply lib/scope.js]: More important than the others (to me) - I'd love an explanation of the different patterns of encoding that different configurations of  variable names can have. For example, how `foo[bar][bash.zoo][boo]` would be double encoded, but not triple encoded (or would it be triple and not quadruple?). I'd rather not lose the information that you so kindly dug up.

File: lib/scope.js [hunk 3]
```diff
@@ -1200,96 +1234,65 @@ module.exports = {
...
+    let checkbox_match = to_handle.match( /^(.*)\[[BR]'(.*)'\]/ );
 
     // Of the format `foo[B'encoded']` or `foo[R'encoded']`
     // See https://github.com/jhpyle/docassemble/blob/595ad0a08572c0188f9ab49b1a6af99a3621cd42/docassemble_base/docassemble/base/standardformatter.py#L2108
 
-    // What about something like foo[R'encoded1'][B'encoded2']? Is that a
-    // thing? Can they be index numbers? Or just `foo['encoded1']`? What
-    // would their format be? The above? `foo[R'actual_choice'][B'encoded']`?
-    // `foo['something'][B'encoded']`?
+    // The B means it's a string, the R means it's 'something else', that was repr'd to a string.
```
Comment (plocket) [inline lib/scope.js]: Not totally crucial, but I think it would be helpful for future me: I think this is generally understandable by context, but not the specifics - `repr` is not a js thing from what I can tell. Is there a js way of saying this? Also, what's an example of 'something else'?

File: lib/scope.js (hunk 2 above)
Comment (plocket) [reply lib/scope.js]: Sorry I'm nitpicking about comments. I just think this is a complex part of the concepts. Maybe the changes I'm suggesting wouldn't even be helpful and I'm just imagining they would. That said, next suggestion:

May be worth adding param descriptions too. You can probably copy a few of them from the function above. Even I'm struggling with it and I originally wrote it.

File: lib/scope.js [hunk 4]
```diff
@@ -1200,96 +1234,65 @@ module.exports = {
...
   getOptionValues: async function ( scope, { $select }) {
     /* Returns all the option node values of the select node. */
     let options = [];
     let option_nodes = $select.find( `option` );
     for ( let opt of option_nodes ) {
       options.push( opt.attribs.value );
-      // Dropdown options created with objects will be encoded
-      let decoded = await scope.fromBase64( scope, { base64_str: opt.attribs.value });
-      options.push( decoded );
+      // Dropdowns created w/ objects are base64 encoded; but it's handled in getPossibleTableRowMatches
```
Comment (plocket) [inline lib/scope.js:1304]: Penny for your thoughts: Does it make sense to make an issue to change that so all the decoding goes on when we get the fields? Or maybe an issue to research that?

File: tests/features/interactive_steps.feature [hunk 5]
```diff
@@ -132,3 +132,17 @@ Scenario: tap element with an error
   """
   Given I start the interview at "test_taps"
   And I tap the "Tests-not_there-tab" tab
+
+@i5 @decoding-test @object-multiselect
+Scenario: Runs through interview designed to use base64 encoding corner cases
```
Comment (plocket) [inline tests/features/interactive_steps.feature]:
```suggestion
Scenario: Base64 encoded corner cases are decoded correctly
```

File: lib/scope.js [hunk 6]
```diff
@@ -1200,96 +1234,65 @@ module.exports = {
...
-          // It may have invalid chars if it's not a one-word var name, though, like '_'
-          let box_layer_2_decoded = await scope.fromBase64( scope, { base64_str: box_layer_1_decoded });
-          var_names.push(`${ var_name }['${ box_layer_2_decoded }']`);
-        } catch ( err ) {}  // It wasn't double encoded. Do nothing.
+      if ( nota ) { 
+        possible.push( `${ var_name }['None']` ); 
+      } else {
+        if (checkbox_match[2].match(da_base64_regex)) {
+          // Get the choice name. E.g. option_1, option_2
+          let box_to_handle = fromBase64( checkbox_match[2] );
```
Comment (plocket) [inline lib/scope.js]: Can we change `box_to_handle` back to `box_layer_1_decoded`? I find the first option too vague and hard to understand what it's doing. The second one seems much more specific to what it's doing and matches the later `box_layer_2_decoded`. Or something that otherwise indicates that we're peeling back layers and this is just the first one? I know the second one may not happen in the end, but it's still the first later, even if it turns out to be the only one.

I do realize I made the terrible variable name `to_handle`, but I wish I hadn't done that.

File: lib/scope.js [hunk 7]
```diff
@@ -54,7 +54,33 @@ let is_secret_var = function(row) {
   return original.flags !== undefined && original.flags.secret_name !== undefined;
 }
 
+let fromBase64 = function(base64_str) {
+  /* Takes a base64 encoded string and returns it decoded.
+   * Note that this accept take strings with invalid base 64 characters,
```
Comment (plocket) [inline lib/scope.js]: Very good stuff to know!

```suggestion
   * Note that this can take strings with invalid base 64 characters,
```

File: lib/scope.js [hunk 8]
```diff
@@ -1200,96 +1234,65 @@ module.exports = {
       // TODO: Should we move this back into `getAllFields()`?
     }
 
-    // A lot of checkbox fields have a key hidden inside them. (not yesno type checkboxes)
-    let checkbox_match = layer_1_decoded.match( /^(.*)\[[BR]'(.*)'\]/ );
+    return scope.nestedDecode(scope, { to_handle: layer_1_decoded, field_like_names, nota });
+  },  // Ends scope.getPossibleVarNames()
+
+  nestedDecode: function ( scope, { to_handle, field_like_names = {}, nota } ) {
```
Comment (plocket) [inline lib/scope.js]: Not a huge thing, but is there a better name for this? If it's decoding nested values, maybe `decodeNested` instead? Some kind of action/verb instead of a noun?

**Review by plocket — APPROVED** [review]
The code looks good! Great changes! Some comments on clarity and some typo corrections, but this is probably good to go as far as functionality is concerned. There is one comment I feel is more important to add - the note about capturing da's encoding patterns.

This is a complex constellation of concepts to tackle. Kudos.

File: lib/scope.js [hunk 9]
```diff
@@ -54,7 +54,33 @@ let is_secret_var = function(row) {
   return original.flags !== undefined && original.flags.secret_name !== undefined;
 }
 
+let fromBase64 = function(base64_str) {
+  /* Takes a base64 encoded string and returns it decoded.
+   * Note that this accept take strings with invalid base 64 characters,
+   * changes those characters to `/`s instead.
```
Comment (BryceStevenWilley) [inline lib/scope.js]:
```suggestion
   * but will change those characters to `/`s instead.
```

File: lib/scope.js (hunk 3 above)
Comment (BryceStevenWilley) [reply lib/scope.js]: I don't know of an equivalent in javascript, but I do want to stick with exact descriptions: docassemble is using the `repr` function in python. I don't think trying to describe what's happening would make sense in javascript, as it's very specific to the python objects internal to docassemble. I can make it clearer that it's talking about python process internal to docassemble though.

Something else is literally just anything that's not a string in docassemble. I'll make that clearer, but I don't have any specific examples. None of this is documented in docassemble, so it's hard to tie my understanding of docassemble internals to concrete examples.

File: lib/scope.js (hunk 1 above)
Comment (BryceStevenWilley) [reply lib/scope.js]: Good catch, I had meant to simplify things by taking away the object wrapper.

File: lib/scope.js (hunk 4 above)
Comment (BryceStevenWilley) [reply lib/scope.js:1304]: Not really. `getPossibleTableRowMatches` is called only a few lines after `getOptions`, so they're very closely tied together anyway. 

If you want to, I could remove `getOptions`; it's only used in one place, and the one calling function, `getAllFields`, already has a lot of other implementation details for other types of fields, like checkboxes, so there's not a lot of benefit of abstracting `getOptions` out.

File: lib/scope.js (hunk 6 above)
Comment (BryceStevenWilley) [reply lib/scope.js]: Agreed, `box_to_handle` is a bad name. It's not too hard to just change the name of `to_handle` to be better, so I'll do that.

The thing I don't quite like about the `layer_1_decoded` part is that this is now a separate function, so the layer numbers aren't actually correct. We've already decoded the whole string once outside of this function, so the numbers don't match my mental model that includes the whole process of decoding. Also, box is a vague term too.

I changed `to_handle` to `expr_str`, `box_to_handle` to `dictionary_key_decoded` (dictionary and not object because it's referring to the python dictionary that we're handling), and removed the need to have a named variable for the `layer_2` decoding, take another look and let me know if that addresses your confusion.

File: lib/scope.js (hunk 8 above)
Comment (BryceStevenWilley) [reply lib/scope.js]: I landed on `decodeFully`, since it might or might not be nested, and "nested" has different meanings to me before vs after "decode".

File: lib/scope.js (hunk 2 above)
Comment (BryceStevenWilley) [reply lib/scope.js]: I added param descriptions, and more comments in the function itself. I also removed some older, conflicting comments that weren't quite accurate with our new knowledge. Let me know if those help.

Comment (plocket) [reply lib/scope.js]: Some great updates! It doesn't quite cover the different variations in the ways that I wanted, but maybe that can be made into its own issue. e.g.:

Here are the different encoding patterns DA creates for variable names:
- `foo` -> "encoded str"
- `foo[bar]` -> "encoded str" -> "foo[B'encoded str]" -> "foo[bar]"
- `foo[bar][baz][zoo]` -> "encoded str" -> "foo[B'encoded str]" -> "foo[B'encoded str]" - > "foo[bar][baz][zoo]"
- So on for the different patterns you found

Here are the different encoding patterns DA creates for values:
- Etc. Though maybe these are the same.

I think it's worth being comprehensive here because the details really make a difference for how we've decided to do this. I would have expected to need a loop, but we don't, and this shows the justification for that.

Again, this can be an issue for the future, referencing this comment: https://github.com/SuffolkLITLab/ALKiln/issues/574#issuecomment-1205562550. We can maybe use the interview you created in order to use legit encodings.

File: lib/scope.js (hunk 3 above)
Comment (plocket) [reply lib/scope.js]: Ah, B means the encoded value started as a string, R is that it started as a non-string, like a number or array or whatever.

File: lib/scope.js (hunk 4 above)
Comment (plocket) [reply lib/scope.js:1304]: I think we'll just leave this for another time. No need to do unrelated refactoring in this PR and not particularly crucial.

File: lib/scope.js (hunk 6 above)
Comment (plocket) [reply lib/scope.js]: The change is better in some ways and I think we can leave it at that. What I appreciated about the previous variable names was that they expressed right there in the name (as opposed to a comment) that we were digging down into multiple decodings. We can wait for the future to tell us if it needs more clarity.

File: lib/scope.js (hunk 2 above)
Comment (BryceStevenWilley) [reply lib/scope.js]: I think pushing it off is the best idea, I don't quite have the time to do the due diligence on all of the variations right now. Making an issue of your most recent comment.

File: lib/scope.js (hunk 3 above)
Comment (BryceStevenWilley) [reply lib/scope.js]: I think it's more like an object (I guess everything in python is an object, but eh), like DAObjects and stuff, yeah.

**Comment by BryceStevenWilley** [comment]
I think all of the comments were addressed and confirmed to be addressed or, (in the instance of #616) made into issues. Going to merge now!
